### PR TITLE
Middleware tests 

### PR DESCRIPTION
**Pull request by mjudeikis**
Simple Middleware tests

@jim-minter wdyt if to add something similar for all Middlewares?

### CODE REVIEW

File: pkg/frontend/middleware/body_test.go [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+		method string
+		header map[string]string
+
+		expectedCode int
+	}{
+		{
+			name:   "Get request - No checking",
+			url:    "/test",
+			method: "GET",
+
```
Comment (asalkeld) [inline pkg/frontend/middleware/body_test.go]: any reason for these newlines? just seems to take up more space

File: pkg/frontend/middleware/body_test.go [hunk 2]
```diff
@@ -0,0 +1,108 @@
+package middleware
+
+import (
+	"bytes"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
```
Comment (jim-minter) [inline pkg/frontend/middleware/body_test.go]: I'm not prepared to bring this library in right now.  We're doing fine in unit tests without it.  If we end up pulling in Ginkgo for e2es, we'll end up pulling in Gomega as well.  I'd rather not use any assertion library.  Also if we decide to use one in unit tests, we should also standardise on using it everywhere.

TL;DR: for now, please just do
```
if err != nil {
	t.Fatal(err)
}
```

File: pkg/frontend/middleware/body_test.go [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+		url    string
+		method string
+		header map[string]string
+
+		expectedCode int
+	}{
+		{
+			name:   "Get request - No checking",
+			url:    "/test",
+			method: "GET",
```
Comment (jim-minter) [inline pkg/frontend/middleware/body_test.go]: http.MethodGet etc

File: pkg/frontend/middleware/body_test.go [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+			expectedCode: 200,
```
Comment (jim-minter) [inline pkg/frontend/middleware/body_test.go]: http.StatusOK etc

File: pkg/frontend/middleware/body_test.go [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+			method: "PATCH",
+			header: map[string]string{
+				"Content-Type": "application/json",
+			},
+
+			expectedCode: 200,
+		},
+	}
+
+	ts := httptest.NewServer(Body(GetTestHandler()))
```
Comment (jim-minter) [inline pkg/frontend/middleware/body_test.go]: didn't know this existed, I wonder if it simplifies my security_test.go.  Will look :-)

File: pkg/frontend/middleware/body_test.go [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+func TestBody(t *testing.T) {
+	assert := assert.New(t)
+
+	tests := []struct {
+		name   string
+		url    string
+		method string
+		header map[string]string
+
+		expectedCode int
```
Comment (jim-minter) [inline pkg/frontend/middleware/body_test.go]: wantStatusCode - we're standardising on want

File: pkg/frontend/middleware/body_test.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+)
+
+func TestBody(t *testing.T) {
+	assert := assert.New(t)
+
+	tests := []struct {
+		name   string
+		url    string
+		method string
+		header map[string]string
```
Comment (jim-minter) [inline pkg/frontend/middleware/body_test.go]: http.Header

File: pkg/frontend/middleware/body_test.go [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+				"Content-Type": "application/json",
+			},
+
+			expectedCode: 200,
+		},
+	}
+
+	ts := httptest.NewServer(Body(GetTestHandler()))
+	defer ts.Close()
+	client := http.Client{}
```
Comment (jim-minter) [inline pkg/frontend/middleware/body_test.go]: use http.DefaultClient below instead

File: pkg/frontend/middleware/body_test.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			var u bytes.Buffer
```
Comment (jim-minter) [inline pkg/frontend/middleware/body_test.go]: I believe this is not needed with httptest?

File: pkg/frontend/middleware/body_test.go [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+			req, err := http.NewRequest(test.method, u.String(), bytes.NewBuffer([]byte("")))
+			assert.NoError(err)
+
+			headers := http.Header{}
+			for n, h := range test.header {
+				headers.Set(n, h)
+			}
+			req.Header = headers
+
+			res, err := client.Do(req)
```
Comment (jim-minter) [inline pkg/frontend/middleware/body_test.go]: resp

File: pkg/frontend/middleware/body_test.go [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+			assert.NoError(err)
```
Comment (jim-minter) [inline pkg/frontend/middleware/body_test.go]: always find a way to make sure you do a resp.Body.Close() after making an http request

File: pkg/frontend/middleware/body_test.go [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+			_, err = ioutil.ReadAll(res.Body)
+			assert.NoError(err)
+			assert.Equal(test.expectedCode, res.StatusCode)
+		})
+
+	}
+}
+
+func GetTestHandler() http.HandlerFunc {
```
Comment (jim-minter) [inline pkg/frontend/middleware/body_test.go]: this just boils down to `http.HandlerFunc(func(http.ResponseWriter, *http.Request) {})`

File: pkg/frontend/middleware/body_test.go [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+			url:    "/test",
+			method: "GET",
+
+			expectedCode: 200,
+		},
+		{
+			name:   "Post request - un-supported media type",
+			url:    "/test",
+			method: "POST",
+			header: map[string]string{
```
Comment (jim-minter) [inline pkg/frontend/middleware/body_test.go]: hint, this will become http.Header{"Content-Type": []string{"test"}},

**Review by jim-minter — CHANGES_REQUESTED** [review]

**Comment by jim-minter** [comment]
> @jim-minter wdyt if to add something similar for all Middlewares?

Yes please - that's why I broke them out into a different package / files :-)

File: pkg/frontend/middleware/body_test.go (hunk 5 above)
Comment (mjudeikis) [reply pkg/frontend/middleware/body_test.go]: I rewrote this even more. more will follow.

File: pkg/frontend/middleware/body_test.go (hunk 8 above)
Comment (mjudeikis) [reply pkg/frontend/middleware/body_test.go]: removed overall

File: pkg/frontend/middleware/body_test.go (hunk 9 above)
Comment (mjudeikis) [reply pkg/frontend/middleware/body_test.go]: removed.

**Comment by jim-minter** [comment]
merged - this was OK, but it could really have been much more rigorous - please see the subsequent fixup commit and consider this when implementing the remaining middleware functions :-)
